packages/material-ui/src/ButtonBase/focusVisible.js

```diff
@@ -1,75 +1,111 @@
-import warning from 'warning';
-import ownerDocument from '../utils/ownerDocument';
+// based on https://github.com/WICG/focus-visible/blob/v4.1.5/src/focus-visible.js
 
-const internal = {
-  focusKeyPressed: false,
-  keyUpEventTimeout: -1,
+let hadKeyboardEvent = true;
+let hadFocusVisibleRecently = false;
+let hadFocusVisibleRecentlyTimeout = null;
+
+const inputTypesWhitelist = {
+  text: true,
+  search: true,
+  url: true,
+  tel: true,
+  email: true,
+  password: true,
+  number: true,
+  date: true,
+  month: true,
+  week: true,
+  time: true,
+  datetime: true,
+  'datetime-local': true,
 };
 
-function findActiveElement(doc) {
-  let activeElement = doc.activeElement;
-  while (activeElement && activeElement.shadowRoot && activeElement.shadowRoot.activeElement) {
-    activeElement = activeElement.shadowRoot.activeElement;
+/**
+ * Computes whether the given element should automatically trigger the
+ * `focus-visible` class being added, i.e. whether it should always match
+ * `:focus-visible` when focused.
+ * @param {Element} node
+ * @return {boolean}
+ */
+function focusTriggersKeyboardModality(node) {
+  const { type, tagName } = node;
+
+  if (tagName === 'INPUT' && inputTypesWhitelist[type] && !node.readOnly) {
+    return true;
+  }
+
+  if (tagName === 'TEXTAREA' && !node.readOnly) {
+    return true;
   }
-  return activeElement;
+
+  if (node.isContentEditable) {
+    return true;
+  }
+
+  return false;
 }
 
-export function detectFocusVisible(instance, element, callback, attempt = 1) {
-  warning(instance.focusVisibleCheckTime, 'Material-UI: missing instance.focusVisibleCheckTime.');
-  warning(
-    instance.focusVisibleMaxCheckTimes,
-    'Material-UI: missing instance.focusVisibleMaxCheckTimes.',
-  );
+function handleKeyDown() {
+  hadKeyboardEvent = true;
+}
 
-  instance.focusVisibleTimeout = setTimeout(() => {
-    const doc = ownerDocument(element);
-    const activeElement = findActiveElement(doc);
+/**
+ * If at any point a user clicks with a pointing device, ensure that we change
+ * the modality away from keyboard.
+ * This avoids the situation where a user presses a key on an already focused
+ * element, and then clicks on a different element, focusing it with a
+ * pointing device, while we still think we're in keyboard modality.
+ * @param {Event} e
+ */
+function handlePointerDown() {
+  hadKeyboardEvent = false;
+}
 
-    if (
-      internal.focusKeyPressed &&
-      (activeElement === element || element.contains(activeElement))
-    ) {
-      callback();
-    } else if (attempt < instance.focusVisibleMaxCheckTimes) {
-      detectFocusVisible(instance, element, callback, attempt + 1);
+function handleVisibilityChange() {
```

Comment:
Does it work without this logic? I would expect the active element to be blur and focus. It should restore the previous `hadKeyboardEvent` value. I suspect we can kill `hadFocusVisibleRecently`. But I haven't tried!

Reply:
We should create some fixtures. This is probably best solved with manual testing.

We can take the demo page from the polyfill and create test instructions. I guess there are some quirks for some browsers where visibilitychange is more robust than blur + document.hasFocus.